File: src/lib/binaryProtocol.ts
// バイナリプロトコルのTypeScript実装

import type { DrawEngineCommand } from './bindings';

export interface ImageHeader {
  width: number;
  height: number;
  format: string;
  compressed: boolean;
  originalSize: number;
  compressedSize?: number;
}

export interface ImageChunk {
  chunkIndex: number;
  totalChunks: number;
  data: Uint8Array;
}

// コマンドタイプの列挙（Rust側と同じ値）
export enum CommandType {
  BeginStroke = 0,
  ContinueStroke = 1,
  EndStroke = 2,
  Clear = 3,
  SetBrush = 4,
  SetActiveLayer = 5,
  CreateLayer = 6,
  DeleteLayer = 7,
}

// 描画コマンドのバイナリ表現
export interface DrawCommandBinary {
  commandType: number;
  data: Uint8Array;
}

// 描画コマンドのバッチ
export interface DrawCommandBatch {
  commands: DrawCommandBinary[];
  timestamp: number;
}

export class BinaryProtocol {
  // DrawEngineCommandをバイナリ形式にエンコード（Rust側のenum形式と一致）
  static encodeDrawCommand(command: DrawEngineCommand): DrawCommandBinary {
    // beginStroke - variant index 0
    if (typeof command === 'object' && 'beginStroke' in command) {
      const { x, y, pressure } = command.beginStroke;
      const buffer = new ArrayBuffer(12); // 3 * 4 bytes for Float32
      const view = new DataView(buffer);
      view.setFloat32(0, x, true); // little-endian
      view.setFloat32(4, y, true);
      view.setFloat32(8, pressure, true);
      return {
        commandType: CommandType.BeginStroke,
        data: new Uint8Array(buffer),
      };
    }
    
    // continueStroke
    if (typeof command === 'object' && 'continueStroke' in command) {
      const { x, y, pressure } = command.continueStroke;
      const buffer = new ArrayBuffer(12); // 3 * 4 bytes for Float32
      const view = new DataView(buffer);
      view.setFloat32(0, x, true); // little-endian
      view.setFloat32(4, y, true);
      view.setFloat32(8, pressure, true);
      return {
        commandType: CommandType.ContinueStroke,
        data: new Uint8Array(buffer),
      };
    }
    
    // endStroke
    if (command === 'endStroke') {
      return {
        commandType: CommandType.EndStroke,
        data: new Uint8Array(0),
      };
    }
    
    // clear
    if (command === 'clear') {
      return {
        commandType: CommandType.Clear,
        data: new Uint8Array(0),
      };
    }
    
    // setBrush
    if (typeof command === 'object' && 'setBrush' in command) {
      const brush = command.setBrush;
      // BrushSettingsのバイナリエンコード（bincode形式）
      // size(4) + opacity(4) + color(16) + brushType(4: enumはRustでu32) + blendMode(4) = 32 bytes
      const buffer = new ArrayBuffer(32);
      const view = new DataView(buffer);
      let offset = 0;
      
      // size: f32
      view.setFloat32(offset, brush.size, true);
      offset += 4;
      
      // opacity: f32
      view.setFloat32(offset, brush.opacity, true);
      offset += 4;
      
      // color: [f32; 4] (RGBA)
      for (let i = 0; i < 4; i++) {
        view.setFloat32(offset, brush.color[i], true);
        offset += 4;
      }
      
      // brushType: enum (bincodeではu32としてエンコード)
      const brushTypeValue = brush.brushType === 'pen' ? 0 : brush.brushType === 'brush' ? 1 : 2; // eraser = 2
      view.setUint32(offset, brushTypeValue, true);
      offset += 4;
      
      // blendMode: enum (bincodeではu32としてエンコード)
      const blendModeValue = 
        brush.blendMode === 'normal' ? 0 :
        brush.blendMode === 'multiply' ? 1 :
        brush.blendMode === 'screen' ? 2 : 3; // overlay = 3
      view.setUint32(offset, blendModeValue, true);
      
      return {
        commandType: CommandType.SetBrush,
        data: new Uint8Array(buffer),
      };
    }
    
    // setActiveLayer
    if (typeof command === 'object' && 'setActiveLayer' in command) {
      const layerIndex = command.setActiveLayer;
      const buffer = new ArrayBuffer(8); // usize is u64 in bincode on 64-bit systems
      const view = new DataView(buffer);
      // bincodeのusizeはu64としてエンコード
      view.setBigUint64(0, BigInt(layerIndex), true);
      return {
        commandType: CommandType.SetActiveLayer,
        data: new Uint8Array(buffer),
      };
    }
    
    // createLayer
    if (command === 'createLayer') {
      return {
        commandType: CommandType.CreateLayer,
        data: new Uint8Array(0),
      };
    }
    
    // deleteLayer
    if (typeof command === 'object' && 'deleteLayer' in command) {
      const layerIndex = command.deleteLayer;
      const buffer = new ArrayBuffer(8); // usize is u64 in bincode on 64-bit systems
      const view = new DataView(buffer);
      view.setBigUint64(0, BigInt(layerIndex), true);
      return {
        commandType: CommandType.DeleteLayer,
        data: new Uint8Array(buffer),
      };
    }
    
    throw new Error(`Unknown command type: ${JSON.stringify(command)}`);
  }
  
  // DrawCommandBinaryをシリアライズ（bincode形式）
  static serializeDrawCommandBinary(commandBinary: DrawCommandBinary): Uint8Array {
    // bincode形式: command_type(1 byte) + data(Vec<u8>)
    // Vec<u8>はbincodeでは長さをu64（8バイト）としてエンコード + データ
    const dataLength = commandBinary.data.length;
    
    const totalSize = 1 + 8 + dataLength; // 1 byte for command_type, 8 bytes for length, then data
    const buffer = new ArrayBuffer(totalSize);
    const view = new DataView(buffer);
    const uint8View = new Uint8Array(buffer);
    
    let offset = 0;
    
    // command_type: u8
    uint8View[offset] = commandBinary.commandType;
    offset += 1;
    
    // data: Vec<u8> - u64 length (little-endian) + data
    view.setBigUint64(offset, BigInt(dataLength), true);
    offset += 8;
    
    uint8View.set(commandBinary.data, offset);
    
    return uint8View;
  }
  
  
  // 複数のDrawEngineCommandをバッチとしてエンコード（bincode互換）
  static encodeCommandBatch(commands: DrawEngineCommand[]): Uint8Array {
    const batch: DrawCommandBatch = {
      commands: commands.map(cmd => this.encodeDrawCommand(cmd)),
      timestamp: Date.now(),
    };
    
    // バッチのシリアライズ（bincodeフォーマット）
    // DrawCommandBatch { commands: Vec<DrawCommandBinary>, timestamp: u64 }
    
    // まず各コマンドをシリアライズ
    const serializedCommands = batch.commands.map(cmd => this.serializeDrawCommandBinary(cmd));
    
    // 全体のサイズを計算
    const totalCommandsSize = serializedCommands.reduce((sum, cmd) => sum + cmd.length, 0);
    const totalSize = 8 + totalCommandsSize + 8; // 8 bytes for Vec length, commands data, 8 bytes for timestamp
    
    const buffer = new ArrayBuffer(totalSize);
    const view = new DataView(buffer);
    const uint8View = new Uint8Array(buffer);
    
    let offset = 0;
    
    // commands: Vec<DrawCommandBinary> - u64で長さ（little-endian）、その後要素
    view.setBigUint64(offset, BigInt(batch.commands.length), true);
    offset += 8;
    
    // 各コマンドを書き込み
    for (const cmdData of serializedCommands) {
      uint8View.set(cmdData, offset);
      offset += cmdData.length;
    }
    
    // timestamp: u64 (little-endian)
    const timestamp = BigInt(batch.timestamp);
    view.setBigUint64(offset, timestamp, true);
    
    // デバッグ情報を出力
    console.log('[BinaryProtocol] Encoded command batch:', {
      commandsCount: batch.commands.length,
      totalSize: uint8View.length,
      timestampOffset: offset,
      timestamp: batch.timestamp,
      firstBytes: Array.from(uint8View.slice(0, Math.min(32, uint8View.length)))
        .map(b => b.toString(16).padStart(2, '0'))
        .join(' ')
    });
    
    return uint8View;
  }
  
  // 単一のDrawEngineCommandをバイナリ形式にエンコード（bincode互換）
  static encodeDrawCommandDirect(command: DrawEngineCommand): Uint8Array {
    const commandBinary = this.encodeDrawCommand(command);
    const encoded = this.serializeDrawCommandBinary(commandBinary);
    
    // デバッグ情報を出力
    console.log('[BinaryProtocol] Encoded single command:', {
      command: command,
      commandType: commandBinary.commandType,
      dataSize: commandBinary.data.length,
      totalSize: encoded.length,
      firstBytes: Array.from(encoded.slice(0, Math.min(32, encoded.length)))
        .map(b => b.toString(16).padStart(2, '0'))
        .join(' ')
    });
    
    return encoded;
  }

  // 画像データのデコード
  static decodeImageData(buffer: ArrayBuffer): {
    header: ImageHeader;
    data: Uint8Array;
  } {
    const view = new DataView(buffer);
    
    // ヘッダー長を読み取り（最初の4バイト）
    const headerLength = view.getUint32(0, true); // little-endian
    
    // ヘッダーのJSON部分をデコード
    const headerBytes = new Uint8Array(buffer, 4, headerLength);
    const headerJson = new TextDecoder().decode(headerBytes);
    const header: ImageHeader = JSON.parse(headerJson);
    
    // 画像データ部分を取得
    const dataOffset = 4 + headerLength;
    const data = new Uint8Array(buffer, dataOffset);
    
    return { header, data };
  }
  
  // LZ4解凍（スタブ実装）
  static async decompressLZ4(compressed: Uint8Array, _originalSize: number): Promise<Uint8Array> {
    // TODO: 実際のLZ4解凍実装を追加
    console.warn('LZ4 decompression not implemented yet, returning compressed data');
    return compressed;
  }
  
  // 画像データを解凍してRGBAデータとして返す
  static async decodeAndDecompressImage(buffer: ArrayBuffer): Promise<{
    width: number;
    height: number;
    data: Uint8Array;
  }> {
    const { header, data } = this.decodeImageData(buffer);
    
    let imageData: Uint8Array;
    if (header.compressed && header.compressedSize) {
      // 圧縮されている場合は解凍
      imageData = await this.decompressLZ4(data, header.originalSize);
    } else {
      imageData = data;
    }
    
    return {
      width: header.width,
      height: header.height,
      data: imageData,
    };
  }
  
  // チャンクデータを結合
  static combineChunks(chunks: ImageChunk[]): Uint8Array {
    // チャンクをインデックス順にソート
    chunks.sort((a, b) => a.chunkIndex - b.chunkIndex);
    
    // 全体のサイズを計算
    const totalSize = chunks.reduce((sum, chunk) => sum + chunk.data.length, 0);
    const combined = new Uint8Array(totalSize);
    
    // チャンクを結合
    let offset = 0;
    for (const chunk of chunks) {
      combined.set(chunk.data, offset);
      offset += chunk.data.length;
    }
    
    return combined;
  }
}

// パフォーマンス計測用のユーティリティ
export class PerformanceMeasure {
  private startTime: number;
  private measurements: Map<string, number> = new Map();
  
  constructor() {
    this.startTime = performance.now();
  }
  
  mark(label: string) {
    const elapsed = performance.now() - this.startTime;
    this.measurements.set(label, elapsed);
  }
  
  getReport(): Record<string, number> {
    const report: Record<string, number> = {};
    let previousTime = 0;
    
    for (const [label, time] of this.measurements) {
      report[label] = time - previousTime;
      previousTime = time;
    }
    
    report['total'] = performance.now() - this.startTime;
    return report;
  }
}

File: src-tauri/src/ipc/binary_protocol.rs
use serde::{Deserialize, Serialize};
use crate::drawing_engine::commands::{DrawEngineCommand, BrushSettings};
use crate::drawing_engine::buffer::BufferManager;

/// バイナリメッセージのラッパー構造体
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinaryMessage {
    /// メッセージタイプ
    pub message_type: String,
    /// バイナリデータ
    pub data: Vec<u8>,
}

/// DrawEngineCommandのバイナリ表現
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DrawCommandBinary {
    pub command_type: u8, // コマンドタイプのインデックス
    pub data: Vec<u8>,    // 各コマンドタイプ固有のデータ
}

impl DrawCommandBinary {
    /// DrawEngineCommandからバイナリ表現に変換
    pub fn from_draw_command(command: &DrawEngineCommand) -> Result<Self, bincode::Error> {
        match command {
            DrawEngineCommand::BeginStroke { x, y, pressure } => {
                #[derive(Serialize)]
                struct BeginStrokeData {
                    x: f32,
                    y: f32,
                    pressure: f32,
                }
                let data = bincode::serialize(&BeginStrokeData {
                    x: *x,
                    y: *y,
                    pressure: *pressure,
                })?;
                Ok(DrawCommandBinary {
                    command_type: 0,
                    data,
                })
            }
            DrawEngineCommand::ContinueStroke { x, y, pressure } => {
                #[derive(Serialize)]
                struct ContinueStrokeData {
                    x: f32,
                    y: f32,
                    pressure: f32,
                }
                let data = bincode::serialize(&ContinueStrokeData {
                    x: *x,
                    y: *y,
                    pressure: *pressure,
                })?;
                Ok(DrawCommandBinary {
                    command_type: 1,
                    data,
                })
            }
            DrawEngineCommand::EndStroke => {
                Ok(DrawCommandBinary {
                    command_type: 2,
                    data: vec![],
                })
            }
            DrawEngineCommand::Clear => {
                Ok(DrawCommandBinary {
                    command_type: 3,
                    data: vec![],
                })
            }
            DrawEngineCommand::SetBrush(brush_settings) => {
                let data = bincode::serialize(brush_settings)?;
                Ok(DrawCommandBinary {
                    command_type: 4,
                    data,
                })
            }
            DrawEngineCommand::SetActiveLayer(layer_index) => {
                let data = bincode::serialize(layer_index)?;
                Ok(DrawCommandBinary {
                    command_type: 5,
                    data,
                })
            }
            DrawEngineCommand::CreateLayer => {
                Ok(DrawCommandBinary {
                    command_type: 6,
                    data: vec![],
                })
            }
            DrawEngineCommand::DeleteLayer(layer_index) => {
                let data = bincode::serialize(layer_index)?;
                Ok(DrawCommandBinary {
                    command_type: 7,
                    data,
                })
            }
        }
    }

    /// バイナリ表現からDrawCommandに変換
    pub fn to_draw_command(&self) -> Result<DrawEngineCommand, bincode::Error> {
        match self.command_type {
            0 => {
                #[derive(Deserialize)]
                struct BeginStrokeData {
                    x: f32,
                    y: f32,
                    pressure: f32,
                }
                let data: BeginStrokeData = bincode::deserialize(&self.data)?;
                Ok(DrawEngineCommand::BeginStroke {
                    x: data.x,
                    y: data.y,
                    pressure: data.pressure,
                })
            }
            1 => {
                #[derive(Deserialize)]
                struct ContinueStrokeData {
                    x: f32,
                    y: f32,
                    pressure: f32,
                }
                let data: ContinueStrokeData = bincode::deserialize(&self.data)?;
                Ok(DrawEngineCommand::ContinueStroke {
                    x: data.x,
                    y: data.y,
                    pressure: data.pressure,
                })
            }
            2 => Ok(DrawEngineCommand::EndStroke),
            3 => Ok(DrawEngineCommand::Clear),
            4 => {
                let brush_settings: BrushSettings = bincode::deserialize(&self.data)?;
                Ok(DrawEngineCommand::SetBrush(brush_settings))
            }
            5 => {
                let layer_index: usize = bincode::deserialize(&self.data)?;
                Ok(DrawEngineCommand::SetActiveLayer(layer_index))
            }
            6 => Ok(DrawEngineCommand::CreateLayer),
            7 => {
                let layer_index: usize = bincode::deserialize(&self.data)?;
                Ok(DrawEngineCommand::DeleteLayer(layer_index))
            }
            _ => Err(bincode::Error::new(bincode::ErrorKind::Custom(
                format!("Unknown command type: {}", self.command_type),
            ))),
        }
    }
}

/// キャンバスのイメージデータ
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageDataBinary {
    pub x: f32,
    pub y: f32,
    pub width: u32,
    pub height: u32,
    pub data: Vec<u8>, // RGBA形式のピクセルデータ
}

/// ストロークパスのバイナリ表現
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StrokePathBinary {
    pub points: Vec<(f32, f32, f32)>, // x, y, pressure
    pub color: [f32; 4],
    pub width: f32,
    pub opacity: f32,
}

/// 描画コマンドのバッチ処理用
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DrawCommandBatch {
    pub commands: Vec<DrawCommandBinary>,
    pub timestamp: u64,
}

impl DrawCommandBatch {
    pub fn new() -> Self {
        Self {
            commands: Vec::new(),
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64,
        }
    }

    pub fn add_command(&mut self, command: DrawEngineCommand) -> Result<(), bincode::Error> {
        let binary_command = DrawCommandBinary::from_draw_command(&command)?;
        self.commands.push(binary_command);
        Ok(())
    }

    pub fn to_commands(&self) -> Result<Vec<DrawEngineCommand>, bincode::Error> {
        self.commands
            .iter()
            .map(|cmd| cmd.to_draw_command())
            .collect()
    }
}

/// コマンドタイプの列挙型
#[repr(u8)]
#[derive(Debug, Clone, Copy)]
pub enum CommandType {
    BeginStroke = 0,
    ContinueStroke = 1,
    EndStroke = 2,
    Clear = 3,
    SetBrush = 4,
    SetActiveLayer = 5,
    CreateLayer = 6,
    DeleteLayer = 7,
}

/// バイナリプロトコルのヘルパー関数
pub struct BinaryProtocol;

impl BinaryProtocol {
    /// DrawEngineCommandをバイナリデータにエンコード
    pub fn encode_draw_command(command: &DrawEngineCommand) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let binary_command = DrawCommandBinary::from_draw_command(command)?;
        Ok(bincode::serialize(&binary_command)?)
    }
    
    /// バイナリデータからDrawCommandとCommandTypeをデコード
    pub fn decode_draw_command(data: &[u8]) -> Result<(CommandType, DrawEngineCommand), Box<dyn std::error::Error>> {
        let binary_command: DrawCommandBinary = bincode::deserialize(data)?;
        let command_type = match binary_command.command_type {
            0 => CommandType::BeginStroke,
            1 => CommandType::ContinueStroke,
            2 => CommandType::EndStroke,
            3 => CommandType::Clear,
            4 => CommandType::SetBrush,
            5 => CommandType::SetActiveLayer,
            6 => CommandType::CreateLayer,
            7 => CommandType::DeleteLayer,
            _ => return Err("Unknown command type".into()),
        };
        let command = binary_command.to_draw_command()?;
        Ok((command_type, command))
    }
    
    /// コマンドバッチをエンコード
    pub fn encode_command_batch(commands: &[DrawEngineCommand]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let mut batch = DrawCommandBatch::new();
        for command in commands {
            batch.add_command(command.clone())?;
        }
        Ok(bincode::serialize(&batch)?)
    }
    
    /// コマンドバッチをデコード
    pub fn decode_command_batch(data: &[u8]) -> Result<Vec<DrawEngineCommand>, Box<dyn std::error::Error>> {
        let batch: DrawCommandBatch = bincode::deserialize(data)?;
        Ok(batch.to_commands()?)
    }
    
    /// 画像データをバイナリフォーマットにエンコード
    pub fn encode_image_data(data: &[u8], width: u32, height: u32) -> Vec<u8> {
        let mut result = Vec::new();
        
        // ヘッダー: width (4 bytes), height (4 bytes), compressed (1 byte = 0)
        result.extend_from_slice(&width.to_le_bytes());
        result.extend_from_slice(&height.to_le_bytes());
        result.push(0); // not compressed
        
        // データ
        result.extend_from_slice(data);
        
        result
    }
    
    /// 圧縮された画像データをエンコード
    pub fn encode_image_data_compressed(data: &[u8], width: u32, height: u32) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let compressed_data = BufferManager::compress(data)?;
        
        let mut result = Vec::new();
        
        // ヘッダー: width (4 bytes), height (4 bytes), compressed (1 byte = 1)
        result.extend_from_slice(&width.to_le_bytes());
        result.extend_from_slice(&height.to_le_bytes());
        result.push(1); // compressed
        
        // 圧縮されたデータ
        result.extend_from_slice(&compressed_data);
        
        Ok(result)
    }
    
    /// RenderResultをバイナリフォーマットにエンコード
    pub fn encode_render_result(canvas_id: &str, image_data: &[u8], width: u32, height: u32, timestamp: u64) -> Vec<u8> {
        let mut result = Vec::new();
        
        // Canvas ID長さ (4 bytes) + Canvas ID
        let id_bytes = canvas_id.as_bytes();
        result.extend_from_slice(&(id_bytes.len() as u32).to_le_bytes());
        result.extend_from_slice(id_bytes);
        
        // タイムスタンプ (8 bytes)
        result.extend_from_slice(&timestamp.to_le_bytes());
        
        // 画像データ
        result.extend_from_slice(&Self::encode_image_data(image_data, width, height));
        
        result
    }
    
    /// 圧縮されたRenderResultをエンコード
    pub fn encode_render_result_compressed(
        canvas_id: &str,
        image_data: &[u8],
        width: u32,
        height: u32,
        timestamp: u64
    ) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let mut result = Vec::new();
        
        // Canvas ID長さ (4 bytes) + Canvas ID
        let id_bytes = canvas_id.as_bytes();
        result.extend_from_slice(&(id_bytes.len() as u32).to_le_bytes());
        result.extend_from_slice(id_bytes);
        
        // タイムスタンプ (8 bytes)
        result.extend_from_slice(&timestamp.to_le_bytes());
        
        // 圧縮された画像データ
        result.extend_from_slice(&Self::encode_image_data_compressed(image_data, width, height)?);
        
        Ok(result)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::drawing_engine::commands::{BrushType, BlendMode};

    #[test]
    fn test_draw_command_binary_conversion() {
        // BeginStrokeコマンドのテスト
        let command = DrawEngineCommand::BeginStroke {
            x: 10.0,
            y: 20.0,
            pressure: 0.5,
        };
        let binary = DrawCommandBinary::from_draw_command(&command).unwrap();
        
        // DrawCommandBinaryのシリアライズをテスト
        let serialized = bincode::serialize(&binary).unwrap();
        println!("DrawCommandBinary serialized: {} bytes", serialized.len());
        println!("Bytes: {:02x?}", serialized);
        
        let restored = binary.to_draw_command().unwrap();
        
        match (&command, &restored) {
            (
                DrawEngineCommand::BeginStroke { x: x1, y: y1, pressure: p1 },
                DrawEngineCommand::BeginStroke { x: x2, y: y2, pressure: p2 }
            ) => {
                assert_eq!(x1, x2);
                assert_eq!(y1, y2);
                assert_eq!(p1, p2);
            }
            _ => panic!("Command type mismatch"),
        }
    }

    #[test]
    fn test_brush_settings_serialization() {
        let brush = BrushSettings {
            size: 10.0,
            opacity: 0.8,
            color: [1.0, 0.0, 0.0, 1.0],
            brush_type: BrushType::Pen,
            blend_mode: BlendMode::Normal,
        };
        
        let command = DrawEngineCommand::SetBrush(brush.clone());
        let binary = DrawCommandBinary::from_draw_command(&command).unwrap();
        let restored = binary.to_draw_command().unwrap();
        
        match restored {
            DrawEngineCommand::SetBrush(restored_brush) => {
                assert_eq!(brush.size, restored_brush.size);
                assert_eq!(brush.opacity, restored_brush.opacity);
                assert_eq!(brush.color, restored_brush.color);
            }
            _ => panic!("Command type mismatch"),
        }
    }

    #[test]
    fn test_command_batch() {
        let mut batch = DrawCommandBatch::new();
        
        batch.add_command(DrawEngineCommand::BeginStroke {
            x: 0.0,
            y: 0.0,
            pressure: 1.0,
        }).unwrap();
        
        batch.add_command(DrawEngineCommand::ContinueStroke {
            x: 10.0,
            y: 10.0,
            pressure: 0.8,
        }).unwrap();
        
        batch.add_command(DrawEngineCommand::EndStroke).unwrap();
        
        let commands = batch.to_commands().unwrap();
        assert_eq!(commands.len(), 3);
        
        // バイナリエンコードをテスト
        let encoded = bincode::serialize(&batch).unwrap();
        println!("Encoded batch size: {} bytes", encoded.len());
        println!("First 32 bytes: {:02x?}", &encoded[..encoded.len().min(32)]);
        
        // デコードをテスト
        let decoded: DrawCommandBatch = bincode::deserialize(&encoded).unwrap();
        assert_eq!(decoded.commands.len(), 3);
    }

    #[test]
    fn test_binary_message_serialization() {
        let message = BinaryMessage {
            message_type: "draw_command".to_string(),
            data: vec![1, 2, 3, 4, 5],
        };
        
        let serialized = bincode::serialize(&message).unwrap();
        let deserialized: BinaryMessage = bincode::deserialize(&serialized).unwrap();
        
        assert_eq!(message.message_type, deserialized.message_type);
        assert_eq!(message.data, deserialized.data);
    }
}